x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java

```diff
@@ -321,6 +332,35 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
         public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
             throw new IllegalArgumentException("[semantic_text] fields do not support sorting, scripting or aggregating");
         }
+
+        public QueryBuilder semanticQuery(InferenceResults inferenceResults, float boost, String queryName) {
+            String nestedFieldPath = getChunksFieldName(name());
+            String inferenceResultsFieldName = getEmbeddingsFieldName(name());
+            QueryBuilder childQueryBuilder;
+
+            if (modelSettings == null) {
+                // No inference results have been indexed yet
+                childQueryBuilder = new MatchNoneQueryBuilder();
+            } else if (inferenceResults instanceof TextExpansionResults textExpansionResults) {
+                // TODO: Use WeightedTokensQueryBuilder
+                var boolQuery = QueryBuilders.boolQuery();
+                for (var weightedToken : textExpansionResults.getWeightedTokens()) {
+                    boolQuery.should(
+                        QueryBuilders.termQuery(inferenceResultsFieldName, weightedToken.token()).boost(weightedToken.weight())
+                    );
+                }
+                boolQuery.minimumShouldMatch(1);
+
+                childQueryBuilder = boolQuery;
+            } else if (inferenceResults instanceof TextEmbeddingResults textEmbeddingResults) {
+                float[] inference = textEmbeddingResults.getInferenceAsFloat();
+                childQueryBuilder = new KnnVectorQueryBuilder(inferenceResultsFieldName, inference, null, null);
+            } else {
+                throw new IllegalArgumentException("Unsupported inference results type [" + inferenceResults.getWriteableName() + "]");
+            }
+
+            return new NestedQueryBuilder(nestedFieldPath, childQueryBuilder, ScoreMode.Total).boost(boost).queryName(queryName);
```

Comment:
Using `ScoreMode.Max` seems more appropriate for dense vectors since we don't return all children. That might be different for sparse vectors but we can introduce a parameter later.

Reply:
I thought about this more and it may be worth taking a second look at this. We will always return the parent doc, which will include all children. 

If we use `ScoreMode.Total`, we will return the total score across all chunks. If we use `ScoreMode.Max`, we will return the score of the best matching chunk. Both are valid, but which one is _more_ valid for the generic use case?

Reply:
Max

Reply:
When we search the graph, its always by nearest right now (`max`), so we really only support `max`.